wagtail/wagtailadmin/forms.py

```diff
@@ -311,3 +314,209 @@ def clean(self):
             self.add_error('expire_at', forms.ValidationError(_('Expiry date/time must be in the future')))
 
         return cleaned_data
+
+
+class CollectionForm(forms.ModelForm):
+    class Meta:
+        model = Collection
+        fields = ('name',)
+
+
+class BaseCollectionMemberForm(forms.ModelForm):
+    """
+    Abstract form handler for editing models that belong to a collection,
+    such as documents and images. These forms are (optionally) instantiated
+    with a 'user' kwarg, and take care of populating the 'collection' field's
+    choices with the collections the user has permission for, as well as
+    hiding the field when only one collection is available.
+
+    Subclasses must define a 'permission_policy' attribute.
+    """
+    def __init__(self, *args, **kwargs):
+        user = kwargs.pop('user', None)
+
+        super(BaseCollectionMemberForm, self).__init__(*args, **kwargs)
+
+        if user is None:
+            self.collections = Collection.objects.all()
+        else:
+            self.collections = (
+                self.permission_policy.collections_user_has_permission_for(user, 'add')
+            )
+
+        if self.instance.pk:
+            # editing an existing document; ensure that the list of available collections
+            # includes its current collection
+            self.collections = (
+                self.collections | Collection.objects.filter(id=self.instance.collection_id)
+            )
+
+        if len(self.collections) == 0:
```

Comment:
I don't think so - if there are >1 collections, then we'll want to iterate over `self.collections` later on, to render the dropdown. So, I think it's more efficient overall to do a full fetch here, rather than doing a `count()` now and a full fetch later.